python/ray/data/datasource/bigquery_datasource.py

```diff
@@ -0,0 +1,185 @@
+            read_tasks.append(read_task)
+
+        return read_tasks
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return None
+
+    def _validate_dataset_table_exist(self, project_id: str, dataset: str) -> None:
+        client = bigquery.Client(project=project_id)
+        dataset_id = dataset.split(".")[0]
+        try:
+            client.get_dataset(dataset_id)
+        except exceptions.NotFound:
+            raise ValueError(
+                "Dataset {} is not found. Please ensure that it exists.".format(
+                    dataset_id
+                )
+            )
+
+        try:
+            client.get_table(dataset)
+        except exceptions.NotFound:
+            raise ValueError(
+                "Table {} is not found. Please ensure that it exists.".format(dataset)
+            )
+
+
+class BigQueryDatasource(Datasource):
+    def create_reader(self, **kwargs) -> Reader:
+        return _BigQueryDatasourceReader(**kwargs)
+
+    def write(
+        self,
+        blocks: List[ObjectRef[Block]],
+        ctx: TaskContext,
+        project_id: str,
+        dataset: str,
+    ) -> WriteResult:
+        def _write_single_block(block: Block, project_id: str, dataset: str):
+            block = BlockAccessor.for_block(block).to_arrow()
+
+            client = bigquery.Client(project=project_id)
+            job_config = bigquery.LoadJobConfig(autodetect=True)
+            job_config.source_format = bigquery.SourceFormat.PARQUET
+            job_config.write_disposition = bigquery.WriteDisposition.WRITE_APPEND
+
+            with tempfile.TemporaryDirectory() as temp_dir:
+                fp = os.path.join(temp_dir, f"block_{uuid.uuid4()}.parquet")
+                pq.write_table(block, fp, compression="SNAPPY")
+
+                retry_cnt = 0
+                while retry_cnt < 10:
+                    with open(fp, "rb") as source_file:
+                        job = client.load_table_from_file(
+                            source_file, dataset, job_config=job_config
+                        )
+                    retry_cnt += 1
+                    try:
+                        logger.info(job.result())
+                        break
+                    except exceptions.Forbidden as e:
+                        logger.info("Rate limit exceeded... Sleeping to try again")
+                        logger.debug(e)
+                        time.sleep(11)
+
+        # Set up datasets to write
+        client = bigquery.Client(project=project_id)
+        dataset_id = dataset.split(".", 1)[0]
+        try:
+            client.create_dataset(f"{project_id}.{dataset_id}", timeout=30)
+            logger.info("Created dataset " + dataset_id)
+        except exceptions.Conflict:
+            logger.info(
+                "Dataset "
+                + dataset_id
+                + " already exists. The table will be overwritten"
+                + " if it already exists."
```

Comment:
Nit: f-strings might be cleaner here.

(Something like this)
```suggestion
                f"Dataset {dataset_id} already exists. "
                "The table will be overwritten if it already exists."
```